src/storage.rs

```diff
@@ -323,13 +360,37 @@ impl TryFrom<&StorageProperties> for capi::StorageProperties {
             (null(), 0)
         };
 
+        // Careful: z needs to live long enough
+        let z = if let Some(metadata) = &value.s3_access_key_id {
+            Some(CString::new(metadata.as_str())?)
+        } else {
+            None
+        };
+        let (access_key_id, bytes_of_access_key_id) = if let Some(ref z) = z {
+            (z.as_ptr(), z.to_bytes_with_nul().len())
+        } else {
+            (null(), 0)
+        };
+
+        // Careful: w needs to live long enough
+        let w = if let Some(metadata) = &value.s3_secret_access_key {
+            Some(CString::new(metadata.as_str())?)
+        } else {
+            None
+        };
+        let (secret_access_key, bytes_of_secret_access_key) = if let Some(ref w) = w {
+            (w.as_ptr(), w.to_bytes_with_nul().len())
+        } else {
+            (null(), 0)
+        };
```

Comment:
- I believe there's enough duplication here to warrant a utility function.
```rs
fn prepare_cstring(metadata: Option<&str>) -> Result<(*const i8, usize), std::ffi::NulError> {
    match metadata {
        Some(value) => {
            let c_string = CString::new(value)?;
            let ptr = c_string.as_ptr();
            let length = c_string.to_bytes_with_nul().len();
            // Note: CString is dropped here, but its pointer and length are valid
            Ok((ptr, length))
        }
        None => Ok((null(), 0)),
    }
}

let (access_key_id, bytes_of_access_key_id) = prepare_cstring(value.s3_access_key_id.as_deref())?;
let (secret_access_key, bytes_of_secret_access_key) = prepare_cstring(value.s3_secret_access_key.as_deref())?;
```

- Also, I'm not sure what the comment `Careful: x needs to live long enough` means. What does it imply to not be careful? How can we write the code so that we don't need to worry about being careful?

Reply:
I don't think this would work, and I think it's related to the `careful` comment. If you construct a new `CString` in `prepare_cstring`, it would drop at the end of the function and the pointer would be invalid. I haven't tried to build this but I expect it would fail to compile due to Rust's lifetime restrictions.